packages/captp/src/captp.js

```diff
@@ -411,22 +570,154 @@ export function makeCapTP(ourId, rawSend, bootstrapObj = undefined, opts = {}) {
+      );
+
+      // Send a "trap" message.
+      lastQuestionID += 1;
+      const questionID = `${ourId}#${lastQuestionID}`;
+
+      // Encode the "method" parameter of the CTP_CALL.
+      let method;
+      switch (implMethod) {
+        case 'get': {
+          const [prop] = implArgs;
+          method = serialize(harden([prop]));
+          break;
+        }
+        case 'applyFunction': {
+          const [args] = implArgs;
+          method = serialize(harden([null, args]));
+          break;
+        }
+        case 'applyMethod': {
+          const [prop, args] = implArgs;
+          method = serialize(harden([prop, args]));
+          break;
+        }
+        default: {
+          assert.fail(X`Internal error; unrecognized implMethod ${implMethod}`);
+        }
+      }
+
+      // Set up the trap call with its identifying information and a way to send
+      // messages over the current CapTP data channel.
+      const [isException, serialized] = trapGuest({
+        implMethod,
+        slot,
+        implArgs,
+        trapToHost: () => {
+          // Send the call metadata over the connection.
+          send({
+            type: 'CTP_CALL',
+            epoch,
+            trap: true, // This is the magic marker.
+            questionID,
+            target: slot,
+            method,
+          });
+
+          // Return an IterationObserver.
+          const makeObserverMethod = observerMethod => (...args) => {
+            send({
+              type: 'CTP_TRAP_ITERATE',
+              epoch,
+              questionID,
+              serialized: serialize(harden([observerMethod, args])),
+            });
+          };
+          return harden({
+            updateState: makeObserverMethod('updateState'),
+            fail: makeObserverMethod('fail'),
+            finish: makeObserverMethod('finish'),
+          });
+        },
+      });
+
+      const value = unserialize(serialized);
+      assert(
+        !isThenable(value),
+        X`Trap() reply cannot be a Thenable; have ${value}`,
+      );
+
+      if (isException) {
+        throw value;
+      }
+      return value;
+    };
+
+    /** @type {TrapImpl} */
+    const trapImpl = {
+      applyFunction: makeTrapImpl('applyFunction'),
+      applyMethod: makeTrapImpl('applyMethod'),
+      get: makeTrapImpl('get'),
+    };
+    harden(trapImpl);
+
+    rets.Trap = makeTrap(trapImpl);
+  }
+
+  return harden(rets);
+};
+
+/**
+ * @typedef {Object} LoopbackOpts
+ * @property {typeof defaultFar} Far
+ */
 
 /**
  * Create an async-isolated channel to an object.
  *
  * @param {string} ourId
- * @returns {{ makeFar<T>(x: T): ERef<T>, makeNear<T>(x: T): ERef<T>  }}
+ * @param {Partial<LoopbackOpts>} [opts]
+ * @returns {{
+ *   makeFar<T>(x: T): ERef<T>,
+ *   makeNear<T>(x: T): ERef<T>,
+ *   makeTrapHandler<T>(x: T): T,
+ *   Trap: Trap
+ * }}
  */
-export function makeLoopback(ourId) {
+export const makeLoopback = (ourId, opts = {}) => {
```

Comment:
Should this be moved into a testing support module? When would it be used in production?

Reply:
Nevermind. We want to be able to take code written correctly for the remote case and also apply it to the local case. (but not resolving until we discuss)

Reply:
Saw your note about production use of loopback. Resolving.